File: common/workflow-core/src/main/scala/org/apache/texera/amber/core/state/StateStorage.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.texera.amber.core.state

import org.apache.texera.amber.core.tuple.{Attribute, AttributeType, Schema, Tuple}

/**
  * Two-column wire/storage format for a [[State]] plus its loop-control
  * `loop_counter`.
  *
  * `loop_counter` is loop bookkeeping owned by the (Python) worker runtime and
  * is intentionally NOT part of [[State]]. Whenever state is materialized (the
  * state storage table) or sent over the wire it is written as its own
  * `loop_counter` column parallel to `content`, so it never enters the user
  * state JSON. Scala operators never produce a non-zero counter, so the Scala
  * write paths emit `0`; this object exists so the bilingual state table schema
  * and tuple layout stay byte-for-byte in sync with the Python `StateStorage`.
  */
object StateStorage {
  val Content = "content"
  val LoopCounter = "loop_counter"

  val schema: Schema = new Schema(
    new Attribute(Content, AttributeType.STRING),
    new Attribute(LoopCounter, AttributeType.LONG)
  )

  def toTuple(state: State, loopCounter: Long): Tuple =
    Tuple
      .builder(schema)
      .addSequentially(Array(state.toJson, Long.box(loopCounter)))
      .build()

  def fromTuple(row: Tuple): (State, Long) =
    (
      State.fromJson(row.getField[String](Content)),
      row.getField[java.lang.Long](LoopCounter).toLong
    )
}

File: amber/src/main/python/core/models/state_storage.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from .schema import Schema
from .state import State
from .tuple import Tuple


class StateStorage:
    """Two-column wire/storage format for a ``State`` and its ``loop_counter``.

    ``loop_counter`` is loop-control bookkeeping owned by the worker runtime,
    not part of the user ``State``. In memory it rides on the ``StateFrame``
    envelope; whenever state is serialized (network) or materialized (the state
    storage table) it is written as its own ``loop_counter`` column parallel to
    ``content`` so it never enters the user state JSON.

    This is the single source of truth for the two-column layout. The Scala
    ``StateStorage`` object must stay byte-for-byte in sync (same column names,
    order, and types), since the same state table is written/read by both.
    """

    CONTENT = "content"
    LOOP_COUNTER = "loop_counter"
    SCHEMA = Schema(raw_schema={CONTENT: "STRING", LOOP_COUNTER: "LONG"})

    @staticmethod
    def to_tuple(state: State, loop_counter: int) -> Tuple:
        return Tuple(
            {
                StateStorage.CONTENT: state.to_json(),
                StateStorage.LOOP_COUNTER: int(loop_counter),
            },
            schema=StateStorage.SCHEMA,
        )

    @staticmethod
    def from_tuple(row: Tuple) -> "tuple[State, int]":
        return (
            State.from_json(row[StateStorage.CONTENT]),
            int(row[StateStorage.LOOP_COUNTER]),
        )
